synapse/crypto/keyring.py

```diff
@@ -252,321 +251,225 @@ def verify_events_for_server(
+                    Codes.UNAUTHORIZED,
+                )
 
-        if key_lookups:
-            run_in_background(self._start_key_lookups, key_lookups)
+            verify_key = key_result.verify_key
+            try:
+                json_object = verify_request.get_json_object()
+                verify_signed_json(
+                    json_object,
+                    verify_request.server_name,
+                    verify_key,
+                )
+            except SignatureVerifyException as e:
+                logger.debug(
+                    "Error verifying signature for %s:%s:%s with key %s: %s",
+                    verify_request.server_name,
+                    verify_key.alg,
+                    verify_key.version,
+                    encode_verify_key_base64(verify_key),
+                    str(e),
+                )
+                raise SynapseError(
+                    401,
+                    "Invalid signature for server %s with key %s:%s: %s"
+                    % (
+                        verify_request.server_name,
+                        verify_key.alg,
+                        verify_key.version,
+                        str(e),
+                    ),
+                    Codes.UNAUTHORIZED,
+                )
 
-        return results
+    async def _inner_fetch_key_requests(
+        self, requests: List[_FetchKeyRequest]
+    ) -> Dict[str, Dict[str, FetchKeyResult]]:
+        """Processing function for the queue of `_FetchKeyRequest`."""
+
+        logger.debug("Starting fetch for %s", requests)
+
+        # First we need to deduplicate requests for the same key. We do this by
+        # taking the *maximum* requested `minimum_valid_until_ts` for each pair
+        # of server name/key ID.
+        server_to_key_to_ts = {}  # type: Dict[str, Dict[str, int]]
+        for request in requests:
+            by_server = server_to_key_to_ts.setdefault(request.server_name, {})
+            for key_id in request.key_ids:
+                existing_ts = by_server.get(key_id)
+
+                if existing_ts:
+                    by_server[key_id] = max(request.minimum_valid_until_ts, existing_ts)
+                else:
+                    by_server[key_id] = request.minimum_valid_until_ts
+
+        deduped_requests = [
+            _FetchKeyRequest(server_name, minimum_valid_ts, [key_id])
+            for server_name, by_server in server_to_key_to_ts.items()
+            for key_id, minimum_valid_ts in by_server.items()
+        ]
+
+        logger.debug("Deduplicated key requests to %s", deduped_requests)
+
+        # For each key we call `_inner_verify_request` which will handle
+        # fetching each key. Note these shouldn't throw if we fail to contact
+        # other servers etc.
+        results_per_request = await yieldable_gather_results(
+            self._inner_fetch_key_request,
+            deduped_requests,
+        )
```

Comment:
I thought that was basically the case now, where we blocked waiting for previous look ups for keys for a server?

I think we could move the deduplication into the key fetchers themselves? But that doesn't really help much since we're batching/queuing up request for keys for a particular server anyway.

Reply:
> I thought that was basically the case now, where we blocked waiting for previous look ups for keys for a server?

huh. yes. Ok, ignore this then.